src/coreclr/jit/emit.cpp

```diff
@@ -7095,8 +7095,23 @@ unsigned emitter::emitEndCodeGen(Compiler*         comp,
     *instrCount                                       = 0;
     jitstd::list<RichIPMapping>::iterator nextMapping = emitComp->genRichIPmappings.begin();
 #endif
+#if defined(DEBUG) && defined(TARGET_ARM64)
+    instrDesc* prevId = nullptr;
+#endif // DEBUG && TARGET_ARM64
+
     for (insGroup* ig = emitIGlist; ig != nullptr; ig = ig->igNext)
     {
+
+#if defined(DEBUG) && defined(TARGET_ARM64)
+        instrDesc* currId = emitFirstInstrDesc(ig->igData);
+        for (unsigned cnt = ig->igInsCnt; cnt > 0; cnt--)
```

Comment:
i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted *last*). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

Reply:
> this will not catch case where `movprfx` was in previous IG and following SVE instruction is in new IG.

That should be ok. `prevId` is defined outside the IG loop and isn't reset ever. So on the next IG, `prevId` should be the last instruction in the previous `IG`.

> i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted _last_). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

I remember trying to do that when we did the peephole optimisations, and there was push back. I can't remember the issues around it. But maybe that is why `EMIT_BACKWARDS_NAVIGATION` was added instead?

Reply:
> prevId is defined outside the IG loop and isn't reset ever

that should be fine then.

> that is why EMIT_BACKWARDS_NAVIGATION was added instead

It was added specifically to track more than 1 last instruction for xarch zero extension peephole IIRC. I realized that we also have `emitRemoveLastInstruction()` so perhaps doing the validation once everything is emitted makes sense.